### DISCUSSION TITLE
Catch-all Interceptor

### DISCUSSION THREAD

**Discussion by schmijos** [cypress-io/cypress#25624] [Questions and Help]
I'm trying to setup a custom catch-all interceptor with [`intercept`](https://docs.cypress.io/api/commands/intercept) like this:

```ts
beforeEach(() => {
  cy.intercept({ url: "**" }, (req) => {
    const url = new URL(req.url);
    if (url.hostname === "localhost") {
      req.continue()
    } else {
      req.destroy()
      throw new Error(`Catch-call interceptor catched ${req.url}`);
    }
  }).as("catchAll")
});
```

I know about the existence of [`blockHosts`](https://docs.cypress.io/guides/references/configuration#blockHosts). But I would like to guide our developers towards using stubs instead of making real requests. So just blocking with an HTTP 503 will not serve this purpose comfortably.

My current solution works so-so, because

* Whitelisted interceptions pass the interceptor but are still logged. I'd rather like to treat those requests as **not** intercepted at all. Maybe there's a possibility to overwrite the `interceptor` command, so that the catch-all interceptor is removed from any matching request?
  ![CleanShot 2023-01-30 at 14 50 33@2x](https://user-images.githubusercontent.com/245443/215495562-e6d2fe6e-d78a-486e-9cf2-840bb6f4a962.png)
* Catched interceptions log an ugly error. I bet, you could do this nicer and attach useful information to the command log. But I didn't find a solution with `cy.log` nor with `Cypress.log` which would allow me not to `throw` an error.
  ![CleanShot 2023-01-30 at 14 54 18@2x](https://user-images.githubusercontent.com/245443/215496493-bb0a2981-4e86-4ad6-8ca3-0aa6a75d3f92.png)
* Leaving away `req.destroy()` gives me actually a solution which is closer to the behavior I'd like because client libraries don't retry their requests. But on the other hand the requests are then not really aborted and the spinners keep spinning.
* Instead of throwing an error, I can also `assert.fail`. I think this is a bit closer to the issue at hand.

Does anyone have a nice solution to those problems?

**Comment by portyu9** [comment]
Hi @schmijos 

You can get much closer to this behavior by taking advantage of Cypress's intercept ordering.

Normal `cy.intercept()` handlers run in **reverse order of definition**. Therefore, define your catch-all first, then define each test's legitimate stubs/allow-list routes afterwards.

For example:

```js
let unexpectedRequests

beforeEach(() => {
  unexpectedRequests = []

  // Registered first => lower priority than intercepts declared later.
  cy.intercept('**', (req) => {
    const url = new URL(req.url)

    if (url.hostname === 'localhost') {
      req.continue()
      return
    }

    unexpectedRequests.push(`${req.method} ${req.url}`)

    // Do not make the real external request.
    req.reply({
      statusCode: 599,
      body: 'Unexpected request from Cypress test',
    })
  })
})

afterEach(() => {
  expect(
    unexpectedRequests,
    'unexpected network requests',
  ).to.deep.equal([])
})
```

Then a test can define its expected requests normally:

```js
cy.intercept('GET', 'https://api.example.com/users', {
  fixture: 'users.json',
}).as('users')
```

Because that stub was registered later, it runs before the catch-all and its static response terminates the request phase.

For an allowed request that should really go to the server, explicitly terminate the request phase with `req.continue()`:

```js
cy.intercept('GET', 'https://api.example.com/health', (req) => {
  req.continue()
})
```

A spy-only intercept that does not reply or continue is different: Cypress will continue to the next matching handler, so the catch-all can still receive it.

I would also remove `.as('catchAll')` if you do not want that alias appearing all over the Command Log. Cypress already logs XHR/fetch traffic by default.

There is currently no supported way to decide `log: false` dynamically from inside the intercept callback. Logging has to be configured when the intercept is created.

Finally, `req.destroy()` deliberately creates a network failure. If your application's HTTP client retries network failures, those retries are application/client behavior and Cypress cannot simultaneously simulate a failed connection and guarantee that the client will not retry it. Returning a deterministic synthetic HTTP response and failing once in `afterEach()` generally produces a much cleaner diagnostic.

Please mark this answer as accepted if it helped, thank you.
